markdown/blog/shai-hulud-postmortem.md

```diff
@@ -0,0 +1,101 @@
+    let _0x3ed61a = process.argv[0x1];
+    if (_0x3ed61a && (await My1(_0x3ed61a))) {
+      let _0x1028dd = await mL0(_0x3ed61a);
+      if (_0x1028dd !== null) {
+        let _0x4cc8b3 = a0_0x459ea5.join(_0x349b3d, "package", "bun_environment.js");
+        await iL0(_0x4cc8b3, _0x1028dd);
+      }
+    }
+  }
+```
+
+This led to [24k compromised repositories](https://github.com/search?q=Sha1-Hulud%3A+The+second+coming&type=repositories) spawning on GitHub.
+
+![Affected Repositories](/img/posts/shai-hulud-postmortem/affected.png)
+
+## Our Response
+
+Upon learning about the incident, we immediately initiated our incident response protocol. Our first steps included:
+- Revoking all npm and OpenVSX tokens to prevent further unauthorized access.
+- Deprecating the affected package versions on npm and OpenVSX registries.
+- Collaborating with the npm and OpenVSX security teams to remove the malicious package versions.
+- Maintaining open communication with our community and security researchers in slack.
+
+### Next Steps
+
+- Conduct a thorough security audit of our publishing processes and infrastructure.
+- Publish a GHSA detailing the incident and our response.
+
+## How it happened: attack chain
+
+The investigation is still ongoing, but preliminary findings suggest that the attacker gained access to our npm and OpenVSX publishing tokens, possibly through a compromised CI/CD pipeline or a leaked token in a public repository. Once they had access, they were able to publish malicious versions of our packages without our knowledge. We don't yet know when or how the initial compromise occurred, but as our CI/CD pipelines have recently been audited, we feel that the token leak had occurred a long time ago as we have not rotated the NPM token.
+
+## Are you affected?
+
+If you have used any of the affected packages, we recommend taking the following steps:
+1. **Audit Your Environment**: Check for any unusual activity or changes in your environment variables
+2. **Update Dependencies**: Ensure that you are using the latest, non-malicious versions of the affected packages.
+3. **Rotate Credentials**: If you suspect that any sensitive information may have been compromised, rotate your credentials immediately.
+4. Review your [GitHub security log](https://github.com/settings/security-log?q=action%3Arepo.create) for suspicious repositories that were created unexpectedly.
+5. Check your ~/.bashrc or ~/.zshrc for suspicious additions like sudo shutdown -h 0.
+
+## Lessons Learned
+
+Regardless of how much we prepare, security incidents can still occur. This incident has highlighted several areas for improvement:
+
+- We no longer plan to use NPM tokens for our publishing process and were in the process of switching to the recently released [Trusted Publisher](https://docs.npmjs.com/trusted-publishers/) using OIDC authentication when this incident occurred 😢. This effectively connects our GitHub repo, our CI pipeline, and the NPM registry.
```

Comment:
Avoid using abbreviations if they had not been fully defined before. Not all reader know what OIDC is and maybe we can do away with the emoji since this is a formal report

Reply:
Okay.